File: EmployeeImp.java
// Time Complexity : O(n) where n is the number of employees (if all emps are used up)
// Space Complexity : O(n) since we have to make a hm for all the emp ids
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/**
 * simple traversal(ordering doesnt matter so it can be done in both dfs/bfs)
 * create a hm with key as emp id and value as the employee object itself add
 * the main root to queue pop the queue and add its importance and its
 * subordinates
 */
class Solution {
    public int getImportance(List<Employee> employees, int id) {
        int res = 0;

        // Creating (adjacency+values) list
        Map<Integer, Employee> hm = new HashMap<>();

        for (Employee emp : employees) {
            hm.put(emp.id, emp);
        }

        if (!hm.containsKey(id))
            return 0;

        Queue<Integer> q = new LinkedList<>();
        q.offer(id);

        while (!q.isEmpty()) {

            int boss = q.remove();
            Employee boss_emp = hm.get(boss);
            res += boss_emp.importance;

            for (int i : boss_emp.subordinates) {
                q.offer(i);
            }

        }
        return res;
    }
}

File: employeeimp.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

# Iterative solution
# Time complexity - O(n)
# Space complexity - O(2n)
# Did this solution run on leetcode? - yes
from collections import deque
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        # create a dictionary of employees with their importance score and subordinates information.  
        # Adjacency list
        empDict = {empInfo.id: empInfo for empInfo in employees}    # space - O(n) # id - storing the object reference / id - index of the object in the array
        
        importance = 0
        q = deque([id])
        while q:    #space - O(n-1)
            currEmp = q.popleft()
            importance += empDict[currEmp].importance   # sum up the importance scores
            q.extend(empDict[currEmp].subordinates)     # add the subordinates information to the queue.
            
        return importance
            

# Recursive solution
# Time complexity - O(n)
# Space complexity - O(2n) # n for the implicit stack and n for the hash map
# Did this solution run on leetcode? - yes
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        # create a dictionary of employees with their importance score and subordinates information.            
        empDict = {empInfo.id: empInfo for empInfo in employees}
        # recursive call.
        def rec(eid):
            return empDict[eid].importance + sum(rec(emp) for emp in empDict[eid].subordinates)
        return rec(id)
        